numcodecs/compat.py

```diff
@@ -27,80 +25,150 @@
-            else:
-                order = 'C'
-            buf = buf.reshape(out.shape, order=order)
+    return out
 
-        # copy via numpy
-        np.copyto(out, buf)
 
-    # handle generic buffer destination
-    else:
+def ensure_text(l, encoding='utf-8'):
+    if isinstance(l, text_type):
+        return l
+    else:  # pragma: py3 no cover
+        return text_type(l, encoding=encoding)
 
-        # obtain memoryview of destination
-        dest = memoryview(out)
 
-        # ensure source is 1D
-        if isinstance(buf, np.ndarray):
-            buf = buf.reshape(-1, order='A')
-            # try to match itemsize
-            dtype = 'u%s' % dest.itemsize
-            buf = buf.view(dtype=dtype)
+def ensure_ndarray_from_memory(o, flatten=True):
+    """Convenience function to obtain a numpy ndarray using memory exposed by object `o`,
+    ensuring that no memory copies are made, and that `o` is not an object array.
 
-        # try to copy via memoryview
-        dest[:] = buf
+    Parameters
+    ----------
+    o : bytes-like
+        Any object exposing a memory buffer. On Python 3 this must be an object exposing
+        the new-style buffer interface. On Python 2 this can also be an object exposing
+        the old-style buffer interface.
+    flatten : bool, optional
+        If True, flatten any multi-dimensional inputs into a one-dimensional memoryview.
 
-    return out
+    Returns
+    -------
+    o : ndarray
 
+    """
 
-def ndarray_from_buffer(buf, dtype):
-    if isinstance(buf, np.ndarray):
-        arr = buf.reshape(-1, order='A').view(dtype)
-    else:
-        arr = np.frombuffer(buf, dtype=dtype)
-    return arr
+    if not isinstance(o, np.ndarray):
 
+        # first try to obtain a memoryview or buffer, needed to ensure that we don't
+        # subsequently copy memory when going via np.array()
 
-def ensure_text(l, encoding='utf-8'):
-    if isinstance(l, text_type):
-        return l
-    else:  # pragma: py3 no cover
-        return text_type(l, encoding=encoding)
+        if PY2:  # pragma: py3 no cover
+            # accept objects exposing either old-style or new-style buffer interface
+            try:
+                o = memoryview(o)
+            except TypeError:
+                o = buffer(o)
+
+        else:  # pragma: py2 no cover
+            o = memoryview(o)
+
+        # N.B., this is not documented, but np.array() will accept an object exposing
+        # a buffer interface, and will take a view of the memory rather than making a
+        # copy, preserving type information where present
+        o = np.array(o, copy=False)
+
+    # check for object arrays
+    if o.dtype == object:
+        raise ValueError('object arrays are not supported')
+
+    if flatten:
+
+        # flatten the array to 1 dimension
+        o = o.reshape(-1, order='A')
+
+    return o
+
+
+def ensure_memoryview(o, flatten=True):
+    """Obtain a :class:`memoryview` with a view of memory exposed by `o`.
```

Comment:
Originally went down this route as well. However what I discovered is that in Python 2 some things only support the old-style buffer protocol and some support the new buffer protocol or both. Admittedly Python 3 is much more straightforward as everything supports the new  buffer protocol where possible.

Given this realization it seemed much better to return something that supported both protocols and was generally more functional (to support reshaping, casting, type checking, etc.). Fortunately NumPy's `ndarray` does all of these things. So it just made more sense to always convert to that. Plus contributors are probably more comfortable with NumPy `ndarray`s than `memoryview`s.

Reply:
My leaning would to write `ensure_memoryview(...)` as `memoryview(ensure_contiguous_ndarray(...))` where possible and try to make sure that works (changing `ensure_contiguous_ndarray` as needed). Expect that users are quite happy to stick with `ndarray`s and only go to a `memoryview` in unusual circumstances. Plus it will be lower cognitive load (even for us 😄).